text/0162-refactoring-support.md

```diff
@@ -0,0 +1,606 @@
+- Add physical ID to resource equivalence.
+
+#### Phase 2 (application)
+
+Once the detection of all cases is implemented in phase 1, we are ready to
+implement the application of the changes.
+
+High-level tasks:
+
+- Add new permissions to the bootstrap stack.
+- Implement the actual refactoring.
+- Implement rollback.
+- Implement the progress bar to display the refactoring progress.
+- Implement feature flags.
+- Handle cross-stack references.
+- Add the refactor step to the `deploy` command.
+- Write a blog post.
+
+### Are there any open issues that need to be addressed later?
+
+This improved deployment experience actually consists of two separate steps,
+behind the scenes: refactoring followed by deployment. And the whole workflow is
+controlled by the CLI. As a result, this is not an atomic operation: it is
+possible that the refactoring step succeeds, but before the CLI has a chance to
+deploy the changes, it gets interrupted (computer crash, network failures, etc.)
+In this case, the user will be left with a stack that is neither in the original
+nor in the desired state.
+
+In particular, the logical ID won't match the CDK construct path, stored in the
+resource's metadata. This has consequences for the CloudFormation console, which
+will show a Tree view that is not consistent with the Flat view.
+
+Possible solutions to consider, from more specific to more general:
+
+- CloudFormation to ignore changes in the `Metadata[aws:cdk:path]` resource
+  attribute in refactor operations.
+- CloudFormation to allow resource additions and deletions in refactor
+  operations.
+- Two-phase commit. The CLI could create the refactor and the changeset, and
+  then have a new command to execute both in a single atomic operation (let's
+  say, a `executeChangeSetAndRefactor()`).
+
+Since all these options depend on changes on the CloudFormation side, and this
+edge case is unlikely to happen, we are going to address it later.
+
+## Appendix
+
+### A. Equivalence between resources
+
+To detect which resources should be refactored, we need to indentify which
+resources have only changed their location, but have remained "the same", in
+some sense. This can be made precise by defining an [equivalence relation] on
+the set of resources.
+
+Before that, let's define a digest function, `d`:
+
+    d(resource) = hash(type + physicalId)                       , if physicalId was defined by the user
+                = hash(type + properties + dependencies.map(d)) , otherwise
+
+where `hash` is a cryptographic hash function. In other words, if a resource has
+a physical ID, we can use the physical ID plus its type to uniquely identify
+that resource. In this case, the digest can be computed from these two fields
+alone. A corollary is that such resources can be renamed and have their
+properties updated at the same time, and still be considered equivalent.
+
+Otherwise, the digest is computed from its type, its own properties (that is,
```

Comment:
Is this CFN type? So refactoring from one CDK construct to another that generated the same CFN type and properties would be possible?

Reply:
Yes to both questions.